docs/API.md

```diff
@@ -0,0 +1,188 @@
+# API
+
+This page describes internal API for creating plugins, contributing, etc.
+
+All methods are under the `spaceship::` namespace. All options starts with `SPACESHIP_`.
+
+TODO: Add an example of custom section.
+
+## `SPACESHIP_VERSION`
+
+An environment variable that defines the version of currently running Spaceship prompt version. Can be used for issue reporting or debugging purposes.
+
+Accessible to any program or script running in a current shell session.
+
+### Example:
+
+```zsh
+echo $SPACESHIP_VERSION
+#> 3.0.0
+```
+
+## `SPACESHIP_ROOT`
+
+> **Attention!** Do not modify the value of this variable! Changing the value may cause the damage of Spaceship installation!
+
+An environment variable that defines the path to Spaceship prompt installation. Spaceship uses this variable for resolving path to sections and utils.
+
+Accessible to any program or script running in a current shell session.
+
+### Example:
+
+```zsh
+echo $SPACESHIP_ROOT
+#> /path/to/spaceship-zsh-theme
+```
+
+## `spaceship::section <color> [prefix] <content> [suffix]`
+
+This command displays prompt section prefixed with `prefix`, suffixed with `suffix` and `content` painted in `color`. **Bold** style is applied by default.
+
+If `SPACESHIP_PROMPT_PREFIXES_SHOW` is `false` or if the section is not the first in the prompt, then `prefix` will be omitted.
+
+If `SPACESHIP_PROMPT_SUFFIXES_SHOW` is `false`, then `suffix` will be omitted.
+
+Both `prefix` and `suffix` are optional. They are equal to empty strings by default.
+
+### Arguments
+
+1. `color` _Required_ — Color for displaying the `content`. Can be any of [basic colors](https://wiki.archlinux.org/index.php/zsh#Colors) or [color codes](https://upload.wikimedia.org/wikipedia/commons/1/15/Xterm_256color_chart.svg).
+2. `prefix` _Optional_ — Prefix before `content`. Usually, it's the value of `SPACESHIP_*_PREFIX`.
+3. `content` _Required_ — The content of the section. Can be any valid value or result of command execution.
+4. `suffix` _Optional_ — Suffix after `content`. Usually, it's the value of `SPACESHIP_*_SUFFIX`.
+
+### Example
+
+```zsh
+# Display prompt section with prefix and suffix
+# Backslash is used to escape line ending
+spaceship::section \
+  "$SPACESHIP_SECTION_COLOR" \
+  "$SPACESHIP_SECTION_PREFIX" \
+  "$SPACESHIP_SECTION_SYMBOL$section_content" \
+  "$SPACESHIP_SECTION_SUFFIX"
+
+# Display prompt section without prefix and suffix
+spaceship::section "$color" "$SPACESHIP_CHAR_SYMBOL"
+```
+
+## `spaceship::exists <command>`
+
+This command validates that given program is available for execution. It returns zero exit code if a `command` exists and non-zero if `command` doesn't.
+
+You can use this utility to check if some program is installed and perform actions conditionally. For example, you can either return an error and exit or continue script's execution.
```

Comment:
Why do we need this instead of `which`? Does `spaceship::exists foo` succeed or fail if `foo` is a function?

Reply:
[Check if a program exists from a Bash script
](https://stackoverflow.com/a/677212/5508862) — here's quite comprehensive explanation why we should not rely on `which`.

Reply:
That is all true for bash, but not for zsh. In zsh:

```sh
% which which
which: shell built-in command
```

I see your implementation is built on top of `command -v`, which I believe will have the same basic behavior as `which` in zsh. I think the current implementation is fine, but it would be nice to document that `spaceship::exists` checks for PATH binaries, functions, and builtins.

Reply:
Done.